optimum/executorchruntime/modeling_executorch.py

```diff
@@ -0,0 +1,234 @@
+"""ExecuTorchModelForXXX classes, allowing to run ExecuTorch Models with ExecuTorch Runtime using the same API as Transformers."""
+
+import logging
+import os
+import warnings
+from pathlib import Path
+from tempfile import TemporaryDirectory
+from typing import TYPE_CHECKING, Any, Dict, List, Optional, Set, Tuple, Union
+
+import torch
+from executorch.extension.pybindings.portable_lib import _load_for_executorch
+from huggingface_hub import hf_hub_download
+from huggingface_hub.constants import HUGGINGFACE_HUB_CACHE
+from huggingface_hub.utils import EntryNotFoundError
+from transformers import (
+    AutoConfig,
+    AutoModel,
+    GenerationMixin,
+    AutoModelForCausalLM,
+    GenerationConfig,
+)
+from transformers.integrations.executorch import TorchExportableModuleWithStaticCache
+from transformers.modeling_outputs import (
+    BaseModelOutput,
+    CausalLMOutput,
+    CausalLMOutputWithPast,
+    ModelOutput,
+)
+
+from ..exporters import TasksManager
+from ..exporters.executorch import main_export
+from ..modeling_base import FROM_PRETRAINED_START_DOCSTRING, OptimizedModel
+
+if TYPE_CHECKING:
+    from transformers import PretrainedConfig
+
+
+logger = logging.getLogger(__name__)
+
+
+class ExecuTorchModelForCausalLM(OptimizedModel):
+    """
+    ExecuTorch model with a causal language modeling head for ExecuTorch Runtime inference.
+    """
+
+    auto_model_class = AutoModelForCausalLM
+
+    def __init__(
+        self,
+        model: "ExecuTorchModule",
+        config: "PretrainedConfig",
+    ):
+        super().__init__(model, config)
+        self.et_model = model
+        print(f"DEBUG all static methods: {self.et_model.method_names()}")
+        self.use_kv_cache = self.et_model.run_method("use_kv_cache")[0]
+        self.max_seq_len = self.et_model.run_method("get_max_seq_len")[0]
+        self.max_batch_size = self.et_model.run_method("get_max_batch_size")[0]
+        self.dtype = self.et_model.run_method("get_dtype")[0]
+        self.bos_token_id = self.et_model.run_method("get_bos_id")[0]
+        self.eos_token_id = self.et_model.run_method("get_eos_id")[0]
+        self.vocab_size = self.et_model.run_method("get_vocab_size")[0]
+
+    def forward(self, input_ids: torch.Tensor, cache_position: torch.Tensor) -> torch.Tensor:
+        return self.et_model.forward((input_ids, cache_position))[0]
+
+    @classmethod
+    def from_pretrained(
```

Comment:
here should be `_from_pretrained` which is used to load the model once exported 


```suggestion
    def _from_pretrained(
```

https://github.com/huggingface/optimum/blob/a7a807c9e712fd9669865358e34c1de072b78d8e/optimum/modeling_base.py#L436

Reply:
@echarlaix I understand that in the base class `_from_pretrained` is the virtual one that needs to be implemented and called in `from_pretrained`. However, I think currently the `from_pretrained` has lots of things that can't be generalized and applied to executorch yet, for example [these code snippet](https://github.com/guangy10/optimum/blob/35eebfe62bf721bbab365f569bd0c73057239732/optimum/modeling_base.py#L383-L434). We want to bypass these code (at least for the initial integration). And `_from_transformers` is going to be deprecated, so essentially I won't need anything in `from_pretrained ` and decide to override it directly in the derived class instead.

Reply:
Updated the code by overriding both `_from_pretrained` and `from_pretrained` to support load from local path and hub (with export).